### DISCUSSION TITLE
Is Conditional Formatting Being Considered for the Future?

### DISCUSSION THREAD

**Discussion by luweiming** [posit-dev/great-tables#493] [Ideas]
Hi All,

Just wondering if conditional formatting is on the road map similar to excel. If x greater than y or more complicated conditions, then bold or changing colors on the fonts, background etc. Thanks.

**Comment by HansMellman** [comment]
Have you tried using .data_color? 

https://posit-dev.github.io/great-tables/reference/GT.data_color.html

![image](https://github.com/user-attachments/assets/10df0254-159a-49e0-8a36-d8b2af225d68)

![image](https://github.com/user-attachments/assets/38803b90-a3ab-4524-a7f1-4ece8ea8569f)

**Reply by luweiming** [reply]
Hi  Hans,

I could get the color gradient work but not based on a calculated value. Do you have some examples? I also want to do bold text based on the same conditions not necessarily with color. I hope this is possible.

**Reply by HansMellman** [reply]
I'm not entirely sure what you mean by a calculated value? Is it stored within the Dataframe that you're trying to pass to GT? The above code example shows the conditional formatting as it would be applied in something like excel where the domain parameter is able to infer the necessary range for the colours to be applied to. In terms of having things in bold text if they're matching a certain condition - you could do a map, that's what I've done for plenty of different aspects of the tables I've created. 

Here's an example that shows how to bold text based on a condition.

Specify a function for the condition:
![image](https://github.com/user-attachments/assets/5324f702-3d5f-4cfe-ada9-1839886c39d3)

Excuse the messy code - take a look at the 'tab_style' near the bottom, where I've commented to highlight it matching up with the map function.

```
## OROY
(
    GT(oroy.head(10))
    .tab_header( # does the below code look better with the vertical align or without it? 
        title=md("<img src='https://i.imgur.com/4uWefKj.png' width='90' height='70' style='vertical-align: middle;'> &#x1F3C8; **Offensive Rookie Of The Year** &#x1F3C8; <img src='https://i.imgur.com/HW5AnCl.png' width='90' height='70' style='vertical-align: middle;'>"), ## add small footballs to the title.
        subtitle=md("_**Final Standings**_")
    )
    .cols_label(
        Total_Points = ("Total"),
        Wkly_Av = ("Average"),
        Team = (" ")
    )
    .tab_spanner(
        label=md('**Points**'),
        columns=['Total_Points', 'Wkly_Av']
    )
    .tab_options(
        container_width = "100%", ## What does this actually do?! 
        table_background_color="#F8F8F8", ## if you use Grey - will need to remove.bg the logos
        heading_background_color="#d50a0a",
        heading_border_bottom_color="black",
        table_border_top_color="#d50a0a",
        column_labels_background_color="#d50a0a",
        column_labels_border_top_color="black",
        column_labels_border_bottom_color="black",
        column_labels_border_top_style="solid",
        # table_body_vlines_style="solid",
        # table_body_hlines_style="solid",
        table_font_names="Calibri",
        table_font_size="20px"
    )
    .cols_align(
        align="center", columns=None
    )
    .fmt_image(
        columns='Team',
        path=img_path,
        file_pattern="{}.png"
    )
    .tab_style(
        style=style.fill(
            color=map_colour),
        locations=loc.body("Position")
        )
    .tab_style(
        style=style.text(color=text_colour),
        locations=loc.body(columns="Position")
    )
    .tab_style(
        style=style.text(weight="bold"),
        locations=loc.body(columns="Name"),
    )
    # .tab_style(
    #     style=style.text(weight="bold"),
    #     locations=loc.body(columns="Rank"),
    # )
    .tab_style(
        style=style.text(weight=weight_map),  # Apply the weight map
        locations=loc.body(columns="Rank")    # Apply to the "Rank" column
    )
    .tab_style(
        style=style.text(style="italic"),
        locations=loc.body(columns="Drafted")
    )
    .tab_style(
        style=[
            style.fill(color="#7AE582"),
            style.text(weight="bold")
        ],
        locations=loc.body(columns=["Total_Points", "Wkly_Av"])) ## Does colouring and bolding actually help?   
)
```    

Then take a look at the result, as you can see numbers that are divisible by 2 are shown as bold - as per the condition specified. 
![image](https://github.com/user-attachments/assets/0283304c-15dc-45c9-a67d-523318ccc70f)

Hopefully that helps!

**Reply by jrycw** [reply]
@HansMellman , thanks for your response in this post—it was very informative! The `.apply()` method works well for simple conditional branches. However, I'd like to introduce you to [`case_when()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.case_when.html#pandas.Series.case_when), a new function in Pandas designed to help users handle more complex conditional branching.
```python
import pandas as pd
from great_tables import GT, style, loc


df_pd = pd.DataFrame({"a": range(1, 6),
                      "b": range(11, 16),
                      "c": range(21, 26)})


def map_color(df_):
    return df_["a"].case_when(
        [
            (lambda s: s <= 2, "grey"),
            (lambda s: (s > 2) & (s <= 4), "green"),
            (lambda s: s > 4, "orange"),
        ]
    )


(
    GT(df_pd).tab_style(
        style=style.fill(map_color),
        locations=loc.body(columns=["a"]),
    )
)
```
![image](https://github.com/user-attachments/assets/a7e7f63b-d304-47c8-b390-b88f287e015c)

**Reply by HansMellman** [reply]
Oh awesome! Thanks so much for sharing @jrycw , that's really useful to know. I will definitely use this going forwards!

**Comment by jrycw** [comment]
It seems there are no follow-up issues related to this discussion, so I'll close it for now.
